extension/src/lib.rs

```diff
@@ -0,0 +1,323 @@
+		// Charge weight for making a call from a contract to the runtime.
+		// `debug_message` weight is a good approximation of the additional overhead of going
+		// from contract layer to substrate layer.
+		// reference: https://github.com/paritytech/ink-examples/blob/b8d2caa52cf4691e0ddd7c919e4462311deb5ad0/psp22-extension/runtime/psp22-extension-example.rs#L236
+		let contract_host_weight = ContractSchedule::<T>::get().host_fn_weights;
+		env.charge_weight(contract_host_weight.debug_message)?;
+
+		let (version, function_id, pallet_index, call_index) = extract_env(&env);
+
+		let result = match FuncId::try_from(function_id) {
+			// Read encoded parameters from buffer and calculate weight for reading `len` bytes`.
+			Ok(function_id) => {
+				// reference: https://github.com/paritytech/polkadot-sdk/blob/117a9433dac88d5ac00c058c9b39c511d47749d2/substrate/frame/contracts/src/wasm/runtime.rs#L267
+				let len = env.in_len();
+				env.charge_weight(contract_host_weight.return_per_byte.saturating_mul(len.into()))?;
+				let params = env.read(len)?;
+				match function_id {
+					FuncId::Dispatch => {
+						dispatch::<T, E, I>(&mut env, version, pallet_index, call_index, params)
+					},
+					FuncId::ReadState => {
+						read_state::<T, E, I>(&mut env, version, pallet_index, call_index, params)
+					},
+				}
+			},
+			Err(e) => Err(e),
+		};
+
+		match result {
+			Ok(_) => Ok(RetVal::Converging(0)),
+			Err(e) => Ok(RetVal::Converging(convert_to_status_code(e, version))),
+		}
+	}
+}
+
+/// Extract (version, function_id, pallet_index, call_index) from the payload bytes.
+fn extract_env<T, E: Ext<T = T>>(env: &Environment<E, BufInBufOutState>) -> (u8, u8, u8, u8) {
+	// Extract version and function_id from first two bytes.
+	let (version, function_id) = {
+		let bytes = env.func_id().to_le_bytes();
+		(bytes[0], bytes[1])
+	};
+	// Extract pallet index and call / key index from last two bytes.
+	let (pallet_index, call_index) = {
+		let bytes = env.ext_id().to_le_bytes();
+		(bytes[0], bytes[1])
+	};
+
+	(version, function_id, pallet_index, call_index)
+}
+
+fn read_state<T: frame_system::Config, E: Ext<T = T>, StateReader: ReadState>(
+	env: &mut Environment<E, BufInBufOutState>,
+	version: u8,
+	pallet_index: u8,
+	call_index: u8,
+	mut params: Vec<u8>,
+) -> Result<(), DispatchError> {
+	const LOG_PREFIX: &str = " read_state |";
+
+	// Prefix params with version, pallet, index to simplify decoding.
+	params.insert(0, version);
+	params.insert(1, pallet_index);
+	params.insert(2, call_index);
+	let (mut encoded_version, mut encoded_read) = (&params[..1], &params[1..]);
+	let version = decode_checked::<VersionedStateRead>(&mut encoded_version)?;
+
+	// Charge weight for doing one storage read.
+	env.charge_weight(T::DbWeight::get().reads(1_u64))?;
+	let result = match version {
+		VersionedStateRead::V0 => {
+			let read = StateReader::decode(&mut encoded_read)?;
+			ensure!(StateReader::contains(&read), UNKNOWN_CALL_ERROR);
```

Comment:
Is this not the same? The call is not included in the runtime to be called by a contract, this can be seen as a call that is unknown to the runtime. Only when the pallet is in the runtime but the calls are not allowed in the chain extension there is reason for calling it filtered. Yet you could still say that the call is unknown to the chain extension